### PR TITLE
WIP: Add Vaultwarden

### LINKED ISSUE #612: Suggestion: Add Password Manager to app store
Reasoning: Bitcoin users need to store passwords for various services. Having passwords accessible from multiple devices is best. Having self-sovereign password manager is even better!

Suggestion: bitwarden_rs is an extremely lightweight implementation of the server-side Bitwarden protocol. There is a docker container available.
Github: https://github.com/dani-garcia/bitwarden_rs

Possible blockers: The official Bitwarden app allows users to put their own server's URL, but does not support tor currently. (iirc Android users could run Orbit and point their Bitwarden app to localhost Orbit and have it forward to the onion URL we set for bitwarden_rs)

But it looks as though they are working on onion URL support.
https://community.bitwarden.com/t/tor-support-for-ios-and-android/12721

### PR DESCRIPTION
**Pull request by junderw**
Closes #612 

**EDIT: bitwarden_rs has changed name to Vaultwarden to avoid trademark issues with Bitwarden.**

# App Submission

### App name
Vaultwarden

### Version
1.19.0

### One line description of the app
Manage your passwords from anywhere.

### Summary of the app
This app is a self-hosted password manager, similar to LastPass or 1Password, but self-sovereign.

Anyone with http access to your umbrel can create an account, and start storing their encrypted passwords on your umbrel.

Everything is encrypted on the client app (Android, iOS, browser extension, etc.), so the server (your umbrel) knows nothing except the user's email address. The worst thing someone with access to your umbrel could do is delete your database, but even then, the client app will retain a copy of the passwords unless you explicitly log out of your device. Client users can also export encrypted backup csv files if they don't trust the availability of your umbrel.

This is perfect for Bitcoin users, it can store your exchange passwords, 2FA QR codes, and everything secret you need kept safe.

You know how to be your own bank. But now, you can be your own synced multi-device password manager, with Vaultwarden!

(Note: Bitwarden and Vaultwarden are not affiliated, but are compatible. So any support issues should be directed at Vaultwarden support channels.)

(Also, consider donating to both Bitwarden and Vaultwarden developers.)

### Developed by
Daniel García

### Developer website
https://github.com/dani-garcia

### Source code repository
https://github.com/dani-garcia/vaultwarden

### Support link
https://vaultwarden.discourse.group/

### Uses
- [ ] Bitcoin Core
- [ ] Electrum server
- [ ] LND

### 256x256 SVG icon
TBD

### App screenshots
TBD

### I have tested my app on:
- [ ] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [ ] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [x] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

### CODE REVIEW

**Review by louneskmt — COMMENTED** [review]
Thanks for working on this @junderw! I don't know if @mayankchhabra and @lukechilds are willing to add non Bitcoin related apps for now, but this is definitely a great idea, especially if there is Tor support.

I would change the description to explain what Bitwarden is, what the user can do with it, etc. Currently, you just explain that Bitwarden RS is a Rust implementation and how the user can donate to the project.

File: apps/bitwarden-rs/docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+x-logging:
+  &default-logging
+  driver: journald
+  options:
+    tag: "umbrel-app {{.Name}}"
+
+services:
+  web:
+    container_name: web
+    image: "bitwardenrs/server@sha256:1cc26a5754dff74dd9df95bbbb79af168cd21dfbd83f627ea72c85fa5852ef15"
```
[the remark this answers is not in the document]
Comment (junderw) [reply apps/bitwarden-rs/docker-compose.yml]: Fixed

**Comment by junderw** [comment]
Fixed issues, added screenshots, updated description.

Regardless of inclusion in umbrel. I highly recommend this to anyone who wants to host their own pw manager.

**Comment by stevenroose** [comment]
@junderw just a heads up that if you don't have explicit individual approval by all the contributors to this repository to publish your modified source code ("Adapted Material"), you are officially in violation of the license of this project and you can be sued:
https://github.com/getumbrel/umbrel/issues/291

**Comment by junderw** [comment]
We cool, right? lol

@mayankchhabra
@nolim1t
@lukechilds
@AaronDewes
@louneskmt
@bguillaumat
@dennisreimann
@goums
@k9ert
@vindard
@apotdevin
@STRML
@gonzaloaune
@PeterXMR

**Comment by STRML** [comment]
We all know that :+1:s on a GitHub comment isn't going to fly legally if a challenge were actually to be brought - better to figure out a license and have existing contributors sign a CLA for contributions to date.

**Comment by nolim1t** [comment]
Ideally something we don't have to sign...

**Comment by dennisreimann** [comment]
> Ideally something we don't have to sign...

Or a license change that makes these inconveniences obsolete? :)

**Comment by nolim1t** [comment]
> > Ideally something we don't have to sign...
> 
> Or a license change that makes these inconveniences obsolete? :)

Yes like this one

https://github.com/getumbrel/umbrel/blob/39c8e3379bf2010a5b2c89685d069d81aadf2eec/LICENSE

**Comment by joemphilips** [comment]
I'm not sure this will help but FYI, there was a similar issue before in rust-lightning.
https://github.com/rust-bitcoin/rust-lightning/issues/659

**Comment by outpoints** [comment]
@junderw do you have a solution for backups? Last time I installed Bitwarden there wasn't a good official way to create backups. Thanks!

**Comment by junderw** [comment]
@outpoints for now, the best way to do backups is through the bitwarden_rs web interface.

1. Login via browser
2. Go to settings
3. Download encrypted backup (the file is encrypted with your master password, so be careful if you change your master password often, as an old backup will only be recoverable with the master password at the time)

**Comment by junderw** [comment]
I know some people that use duplicati to perform automated backups on the server side, but tbh, server side backups can be corrupted if you run the backup job without stopping bitwarden_rs first... which is complicated.

You need a complex cron job or some over-arching app that controls both duplicati and bitwarden_rs. (Maybe someday umbrel could do that, but imo periodically doing manual backup is fine)

**Comment by junderw** [comment]
Bitwarden RS is rebranding to avoid trademark issues with Bitwarden.

**Comment by AaronDewes** [comment]
There asre some merge conflicts. Can you resolve them or rebase this?

**Comment by junderw** [comment]
Fixed merge conflicts... which were all "Porque no los dos?" problems.

Anywho, something I noticed:

1. The web interface uses subtle crypto, which doesn't work over http, doesn't work over onion address on Brave's tor window... So the only way to access the web service is:
  A. set up https reverse proxy (boo)
  B. Only use the Android App + Orbot VPN mode to proxy to the onion hidden service (since the app doesn't have those restrictions.)

Some of the major functions of an account can only be accessed via web portal... so not sure who to talk to about fixing this...

I guess browsers need to implement subtle crypto for onion services... since it's secure.

**Comment by junderw** [comment]
Also, considering they are moving the name away from Bitwarden, but they still use the GPL web interface (which says Bitwarden plastered all over it) I am wondering how to go about the icon and the screenshots.

**Comment by AaronDewes** [comment]
> I guess browsers need to implement subtle crypto for onion services... since it's secure.

I hope they'll do.

> Also, considering they are moving the name away from Bitwarden, but they still use the GPL web interface (which says Bitwarden plastered all over it) I am wondering how to go about the icon and the screenshots.

:man_shrugging: Maybe they'll move away from it in the future?

Also, I'm not sure how soon this can be merged, because AFAIK, Mayank and Luke are focusing on apps related to Bitcoin right now.

**Comment by junderw** [comment]
The subtle crypto issue (not being able to use it on onion services) is the same issue as #355 

Just have to wait for Brave / Tor Browser etc. to treat onion services as "Secure Origins"

**Comment by junderw** [comment]
Actually it seems like Tor Browser (PC) works fine.

Websocket notifications are borked though.

**Comment by junderw** [comment]
Closed in favor of #888

**Pull request closed without merging**